docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+There are no discriminator bits required but the 'as-if' rules triggers as if there were a single bit for determining which end of the space should be used and the remaining values are stored at the end of the space that is strongly aligned: bits 0-31.
+
+On 64-bit systems, the same rules apply, packing the used bits of the pointer into bits 0-50.
+
+On CHERI systems, the pointer types are larger than the `Float64` and so the payload is a 128-bit value, distinguished by the tag bit.
+
+### Example: `T* | None`
+
+In this example, `None` is a singleton type with no fields, used as a placeholder indicating a not-present value.
+The `T*` requires one pointer's worth of space for the payload (as above, this varies depending on the target) and one discriminator state.
+The `None` value requires one discriminator state and no payload.
+
+The largest value is a pointer and so on CHERI systems the tag bit is used as the discriminator, on other architectures the low bit of the pointer is used (because the pointer is the largest type and the discriminator fits in a three-bit value).
+The discriminator values are assigned starting from the singletons, so zero is used for `None` and one for `T*`.
+
+A pattern match expression on this type is compiled into a compare-and-branch-on-{non}-zero, which is a single instruction on most architectures.
+Accessing the `T*` on a CHERI system requires no further operations.
+On integer-pointer systems, it requires subtracting one.
+On any mainstream architecture other than RISC-V, this subtraction can be folded into the immediate offset on a load instruction and so requires no additional instructions.
+
+If the `None` value is used (rather than just being pattern matched on and ignored) then the compiler must be able to materialise the address (or capability) to the singleton.
+This is no worse than if we stored the address inline; however, because in that case the match would require a comparison against the address and so the address would need to be loaded before the pattern match, rather than conditionally in one path.
+
+Objects
+-------
+
+Verona objects contain a header with metadata describing the object, followed by user-defined fields.
+
+### Object headers
+
+Object headers are two pointer-size words.
+These are described in detail in [src/rt/object/object.h](../src/rt/object/object.h).
+
+The first word contains the region metadata.
+This is a discriminated union where the low three bits describe how the remainder is used (for example, as a reference count for one immutable object in an SCC, a pointer to that object for other immutable objects in an SCC, and a pointer to the region metadata for sentinel objects).
+
+The second is a pointer to the descriptor for the object.
+The low bits of this are used by the runtime for memory management state and so must be zeroed before accesses.
+
+The object descriptor layout is still likely to change, the current layout is:
+
+64-bit integer size.
```

Comment:
I'm confused. Does "64-bit integer size" means there's a 64-bit padding before the trace pointer? Or that integers here are 64-bit, but actually all elements are pointers? Or that all pointers are encoded in a 64-bit integer by means of offsets into other known vtables?